Created a REST API that other services can make requests to in order to create notifications

ibnetahir1 opened the pull request:
- Created NotificationController in WebApi project that has 3 action methods for creating a comment/like/follow notification
- Other backend services should access this API in order to create their respective notifications and send those to the clients. This removes the need for the frontend to send notification requests.

Josh0224 committed:
created a controller to communicate with hub

jpw547 committed:
Merge pull request #60 from 2011-fakebook-project3/main

Updating our feature branch

Josh0224 committed:
added code to make tests pass

ibnetahir1 committed:
Created a NotificationControllerTest in Testing project
- The test tests the three controller methods
- They currently fail
- There are 6 lines that need to be updated once everything is merged together for the tests to work

jpw547 committed:
Merge pull request #61 from 2011-fakebook-project3/TDDFail

Created a NotificationControllerTest in Testing project

jpw547 committed:
changed the few lines Hamza recommended in his last commit

jpw547 committed:
FakebookNotifications/FakebookNotifications.Testing/

Josh0224 committed:
team effort to get passing tests. test passed

ibnetahir1 committed:
Merge branch 'main' into feature/api

ibnetahir1 committed:
Merge branch 'feature/api' into TDDPass

jpw547 committed:
Merge pull request #62 from 2011-fakebook-project3/TDDPass

Implemented Controller methods to pass the tests and fixed tests

Josh0224 committed:
added proper returns for API ActionMethods

ibnetahir1 committed:
Refactored NotificationController to remove duplicate code
- A new helper method in the controller to handle the notification creation
- Flipped the description of loggedInUser and triggerUser. triggerUser creates the event, and loggedInUser gets sent the notification
- Changed 1 line in the code to match the changes in the description

ibnetahir1 committed:
Fixed some test in NotificationRepoTest
- DeleteNotification_RepoTest was dependent on CreateNotificationTest to run first
- Changed it so it is not dependent on the other test to run first
- A previous commit with this commit message was also trying to do this, but was reverted

ibnetahir1 committed:
Fixed more test dependencies
- Changed the connected DB on NotificationControllerTest from local db to an actual one
- Fixed OnDisconnectedVerify being affected by a test in a different file

jpw547-zz reviewed (commented):
The tests look alright to me. Quality gate seems bugged...

ibnetahir1 committed:
Addressed code smells from sonarcloud
- Removed unnecessary using statements and variables

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs:
```diff
@@ -0,0 +1,188 @@
+            Mock<IOptions<NotificationsDatabaseSettings>> mockSettings = new();
+            settings = new NotificationsDatabaseSettings
+            {
+                ConnectionString = "mongodb+srv://ryan:1234@fakebook.r8oce.mongodb.net/Notifications?retryWrites=true&w=majority",
```
don't send your database connection string to github, even your test database. remove this string & change the password on it.
if we want a test database, better to run it with docker in the pipeline. we'd want a "trait" on all tests that required that, so we could skip running those tests when we don't want to run the test db.
but you're confused on what you're doing here. if you're mocking the mongodb driver, writing a unit test, then there is zero purpose to having a real database. if you're writing an integration test with a real db, then there's no purpose to mocking mongodb driver. figure out what kind of tests you want, and if you want both kinds, put them in separate namespaces at least.

ibnetahir1 committed:
Fixed sonarcloud code smell
- removed commented out code

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs:
```diff
@@ -0,0 +1,188 @@
+                NotificationsCollection = "Notifications"
+            };
+            mockSettings.Setup(s => s.Value).Returns(settings);
+            NotificationsContext mockDbContext = new(mockSettings.Object, _logger);
```
it's not a mock if you're calling the real constructor. figure out what you're trying to do

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs:
```diff
@@ -0,0 +1,188 @@
+        private readonly UserRepo _userRepo;
+        private readonly NotificationsRepo _noteRepo;
+        private readonly NotificationsDatabaseSettings settings;
+        private readonly NullLogger<NotificationsContext> _logger;
```
use a consistent naming convention for private fields - in this project, that's an initial underscore

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs:
```diff
@@ -0,0 +1,188 @@
+
+            // mocking Mongo db
+            NotificationsRepo noteRepo = new(mockDbContext);
+            UserRepo userRepo = new(mockDbContext, noteRepo);
```
the integration test includes these objects too? yeah, i guess you could write an integration test that spans all those classes. but it's certainly not mocking.

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:
```diff
@@ -0,0 +1,117 @@
+        private readonly INotificationsRepo notificationsRepo;
+        private readonly IUserRepo userRepo;
+
+        public NotificationController(IHubContext<NotificationHub> _hub, INotificationsRepo _notificationsRepo, IUserRepo _userRepo) 
```
inverted naming convention. parameters are camelCase, private fields are _camelCase

escalonn replied:
previous review comment unaddressed.

ibnetahir1 committed:
Addressed bugs from sonarcloud
- Changed return type to Task from void in NotificationHubTest for all tests

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:
```diff
@@ -0,0 +1,117 @@
+                await hub.Clients.Group(user.Email).SendAsync(methodName, notification);
+                return Ok();
+            }
+            catch (Exception e)
```
not every exception is the fault of the client. unexpected exceptions are generally the server's fault.

this exception handling is actively bad in worse ways though: you're preventing the default logging of uncaught exceptions, so it's harder for the devs to detect an issue, while sending the exception to the client, so it's easy for hackers to detect an issue.

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:
```diff
@@ -0,0 +1,117 @@
+                    await hub.Groups.AddToGroupAsync(connection, user.Email);
+                }
+
+                await hub.Clients.Group(user.Email).SendAsync(methodName, notification);
```
i don't know signalr but are you sure this is right? you have to rebuild a group out of a user's connections every time just to send him something? and you have to do it slowly, one by one? sometimes it's unavoidable but you usually want to avoid `await` inside a loop at all costs

ibnetahir1 replied:
The way it is currently set up, we have to do it this way. This sends the notification to all different places the same user is connected from (laptop, mobile, etc.).

escalonn replied:
@ibnetahir1 at least add some comments giving some context that would help future development work to possibly improve this aspect. maybe reference links if you based your code/design on something you read online.

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Hubs/NotificationHub.cs:
```diff
@@ -15,7 +15,7 @@ namespace FakebookNotifications.WebApi.Hubs
     public class NotificationHub : Hub
     {
-
+        public static IHubContext<NotificationHub> Current { get; set; }
```
what's this for?

ibnetahir1 replied:
That was a mistake

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Startup.cs:
```diff
@@ -34,6 +34,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddScoped<INotificationsContext, NotificationsContext>();
             services.AddScoped<IUserRepo, UserRepo>();
             services.AddScoped<INotificationsRepo, NotificationsRepo>();
+            
```
unnecessary diff

escalonn replied:
previous review comment unaddressed.

ibnetahir1 replied:
I believe we were trying to add some things, but then deleted the changes so that one extra line stayed. I guess the right call would have been to restore the file.

escalonn reviewed (changes requested):
fyi, async void is ordinarily bad (look up why). for xunit test methods though, it doesn't matter, so you could have marked as false positive. on the other hand, if we keep consistently to async Task, there are no downsides and future code won't have to manually review those issues.

ibnetahir1 committed:
Addressed changes requested by reviewer
- Fixed NotificationControllerTest class so that it correctly uses Moq to unit test
- In the NotificationController, there is now an initial check to see if parameter values are null or not

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs:
```diff
@@ -0,0 +1,125 @@
+        private readonly Mock<IGroupManager> _mockGroups = new();
+        private readonly Mock<IClientProxy> _mockClientProxy = new();
+        private readonly Mock<IUserRepo> _userRepo = new();
+        private readonly Mock<INotificationsRepo> _noteRepo = new();
```
name these mocks consistently.

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs:
```diff
@@ -0,0 +1,125 @@
+using System.Threading.Tasks;
+using Xunit;
+
+namespace FakebookNotifications.Testing.IntegrationTests
```
these are unit tests, not integration tests, so move them to a different namespace & folder

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs:
```diff
@@ -0,0 +1,125 @@
+            Assert.NotNull(result);
+            _noteRepo.Verify(c => c.CreateNotificationAsync(It.IsAny<Domain.Models.Notification>()), Times.Once);
+            _userRepo.Verify(c => c.GetUserAsync(It.IsAny<string>()), Times.Once);
+            _mockGroups.Verify(c => c.AddToGroupAsync(It.IsAny<string>(), It.IsAny<string>(), new CancellationToken()), Times.Exactly(3));
```
why are you passing in a cancellationtoken in these tests

ibnetahir1 replied:
It would not allow me to use the verify function unless I passed something for all parameters.

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationControllerTest.cs:
```diff
@@ -0,0 +1,125 @@
+            _userRepo.Verify(c => c.GetUserAsync(It.IsAny<string>()), Times.Once);
+            _mockGroups.Verify(c => c.AddToGroupAsync(It.IsAny<string>(), It.IsAny<string>(), new CancellationToken()), Times.Exactly(3));
+            Assert.IsAssignableFrom<IActionResult>(result);
+            Assert.IsType<OkResult>(result);
```
these three assertions on result are redundant in a few ways. the compiler can already guarantee that it's an iactionresult so that call does nothing. and the final IsType call makes the IsAssignableFrom and NotNull calls redundant too.
also, to follow liskov substitution principle, don't require the exact type. so assert `Assert.IsAssignableFrom<OkResult>(result);`, and it replaces the other assertions too. likewise for the similar tests.

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:
```diff
@@ -0,0 +1,116 @@
+        [HttpPost("comment")]
+        public async Task<IActionResult> CommentNotificationAsync(string loggedInUser, string triggerUser, int postId)
+        {
+            return await CreateNotificationAsync(loggedInUser, triggerUser, postId, "comment");
```
at least use constants, not magic strings

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:
```diff
@@ -0,0 +1,116 @@
+            else if (notificationType == "follow")
+            {
+                methodName = "FollowNotification";
+            }
```
no exception if you pass an invalid argument? not good for maintainability.

escalonn commented on FakebookNotifications/FakebookNotifications.Testing/IntegrationTests/NotificationRepoTests.cs:
```diff
@@ -132,12 +132,29 @@ public async Task DeleteNotification_RepoTest()
+
+            //Assert
+            Assert.True(result);
+
```
why are we adding this? at a glance i guess this is an improvement on the existing test... not sure how it was passing before. at least change the test name and the arrange/act/assert comments to reflect what's actually going on.

these tests should be refactored eventually

ibnetahir1 replied:
I removed the unnecessary check.

escalonn commented:
also click 'resolve conversation' on review comments when you push commits that you think resolve them

and just in case this comment gets lost among the older items on this PR - https://github.com/2011-fakebook-project3/notifications/pull/63#discussion_r611923924

ibnetahir1 committed:
Addressed changes requested by reviewer
- consistent naming convention
- reduced redundancy in tests
- fixed logic in NotificationController

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:
```diff
@@ -99,17 +102,22 @@ private async Task<IActionResult> CreateNotificationAsync(string loggedInUser, s
             }
+            else
+            {
+                throw new ArgumentException("notificationType is not valid");
```
use the paramName argument to this constructor as well

escalonn replied:
and use the new "switch expression" syntax, it's a good fit here

escalonn replied:
"use the paramName argument to this constructor as well"
previous review comment not addressed

escalonn commented on FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:
```diff
@@ -34,7 +34,8 @@ public NotificationController(IHubContext<NotificationHub> _hub, INotificationsR
         public async Task<IActionResult> CommentNotificationAsync(string loggedInUser, string triggerUser, int postId)
         {
-            return await CreateNotificationAsync(loggedInUser, triggerUser, postId, "comment");
+            string notificationType = "comment";
```
I meant declare the string "comment" once, and reference that one location both from methods like this and from the CreateNotificationAsync method. that will let the compiler catch typos in the notification type. i guess an enum would be better - you can make it a private nested enum

ibnetahir1 committed:
Addressed changes requested by reviewer
- Created a private NotificationType enum in NotificationController
- used a switch expression to assign methodName

ibnetahir1 commented:
@escalonn I did address that change. The throw statement is moved to the switch expression and it does use the paramName argument.

ibnetahir1 commented on FakebookNotifications/FakebookNotifications.WebApi/Controller/NotificationController.cs:
```diff
@@ -83,30 +87,19 @@ private async Task<IActionResult> CreateNotificationAsync(string loggedInUser, s
-
+                NotificationType.Comment => "CommentNotification",
+                NotificationType.Like => "LikeNotification",
+                NotificationType.Follow => "FollowNotification",
+                _ => throw new ArgumentException("notificationType is not valid", nameof(notificationType))
+            };
```
this is where the throw statement is now.

escalonn reviewed (approved):
you're right, i had the review interface set wrong
